### PR TITLE
Executor quality of life and instrumentation

### PR DESCRIPTION
**Pull request by carterkozak**
**Goals (and why)**:

Preparation for #4701 This will provide additional data to discover and remediate thread bloat.

**Implementation Description (bullets)**:

* PTExecutors thread factory uses daemon threads by default. Most consumers prefer daemon threads to avoid propping up JVMs that want to die.
* PTExecutors.newCachedThreadPool returns ExecutorService instances instead of ThreadPoolExecutor concrete types. This is an ABI break. Please use WC executor factories!
* Instrument NamedThreadFactory using the singleton TaggedMetricRegistry instance to collect data describing thread bloat.

**Testing (What was existing testing like?  What have you done to improve it?)**:

Existing test coverage is sufficient.

**Concerns (what feedback would you like?)**:

¯_(ツ)_/¯

### CODE REVIEW

File: lock-impl/src/test/java/com/palantir/lock/impl/ClientAwareLockTest.java [hunk 1]
```diff
@@ -48,8 +47,7 @@
 @ShouldRetry
 public final class ClientAwareLockTest {
 
-    private static final ExecutorService executor = PTExecutors.newCachedThreadPool(
-            new NamedThreadFactory(ClientAwareLockTest.class.getName(), true));
+    private static final ExecutorService executor = PTExecutors.newCachedThreadPool("ClientAwareLockTest");
```
Comment (jeremyk-91) [inline lock-impl/src/test/java/com/palantir/lock/impl/ClientAwareLockTest.java:50]: nit: Don't feel strongly about which way this falls, but we should probably be consistent on whether we prefer the string literal or `ThisClass.class.getName()` for these executors.

File: changelog/@unreleased/pr-4707.v2.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+type: improvement
```
Comment (jeremyk-91) [inline changelog/@unreleased/pr-4707.v2.yml:1]: We could mark this as `break` if we want to make the ABI break more obvious.

Comment (carterkozak) [reply changelog/@unreleased/pr-4707.v2.yml:1]: Good call, version numbers are cheap :-)

File: lock-impl/src/test/java/com/palantir/lock/impl/ClientAwareLockTest.java (hunk 1 above)
Comment (carterkozak) [reply lock-impl/src/test/java/com/palantir/lock/impl/ClientAwareLockTest.java:50]: that's fair, I'll revert these out to use class.getName(), no reason to churn on these in this change.

**Comment by carterkozak** [comment]
Thanks @jeremyk-91!

**Pull request merged**